capnproto 0.7.0 fails to build on gcc10 on i686 and armv7hl

When rebuilding capnproto 0.7.0 with GCC 10 in Fedora rawhide, I've encountered different failures on i686 and armv7hl.

### i686

The `kj-heavy-tests-run` test suite failed with the following error:
```
2: [ TEST ] filesystem-disk-test.c++:833: DiskFile holes
2: kj/filesystem-disk-test.c++:887: failed: expected file->stat().spaceUsed == meta.spaceUsed
2: stack: 56692dfc 566dd63b f7de4b8b f7d7e441 f7de6388 f7de6aee f7d8fe37 f7d906a5 f7d7e441 f7d8b618 f7de38da f7996ebd 5663b154
2: kj/filesystem-disk-test.c++:920: failed: expected file->stat().spaceUsed < meta.spaceUsed
2: stack: 566d29bc 566dd83a f7de4b8b f7d7e441 f7de6388 f7de6aee f7d8fe37 f7d906a5 f7d7e441 f7d8b618 f7de38da f7996ebd 5663b154
2: [ FAIL ] filesystem-disk-test.c++:833: DiskFile holes (5662 μs)
```
Full build log: [capnproto-c++-0.7.0-5.fc33.i686-build.log.txt](https://github.com/capnproto/capnproto/files/4242628/capnproto-c%2B%2B-0.7.0-5.fc33.i686-build.log.txt)

### armv7hl

The `kj-tests-run` test suite failed with the following error:
```
1: [ TEST ] table-test.c++:786: large tree table
1: *** Received signal #11: Segmentation fault
1: stack: 
1/5 Test #1: kj-tests-run .....................***Failed    1.29 sec
```
Full build log: [capnproto-c++-0.7.0-5.fc33.armv7hl-build.log.txt](https://github.com/capnproto/capnproto/files/4242626/capnproto-c%2B%2B-0.7.0-5.fc33.armv7hl-build.log.txt)

@jakubjelinek has indicated he thinks that capnproto might be buggy here. Quoted from  [RHBZ#1807872#c1](https://bugzilla.redhat.com/show_bug.cgi?id=1807872#c1):

> Tried building it on x86_64 with `-O0 -g -fsanitize=undefined,address` and it FAILs there too, e.g.
```
src/capnp/compiler/module-loader.c++:118:65: runtime error: signed integer overflow: 1535247712000000000 * 73 cannot be represented in type 'long int'
src/kj/array.h:466:14: runtime error: reference binding to misaligned address 0x7f9ca0e74dae for type 'const struct ArrayDisposer', which requires 8 byte alignment
src/kj/string.c++:190:25: runtime error: negation of -9223372036854775808 cannot be represented in type 'long int'; cast to an unsigned type to negate this value to itself
src/kj/array.h:466:14: runtime error: reference binding to misaligned address 0x000000000002 for type 'const struct ArrayDisposer', which requires 8 byte alignment
0x000000000002: note: pointer points here
<memory cannot be printed>
```

> (just a couple of different messages from src/capnp/compiler/capnp-test.sh.log).
> Furthermore, lots of warnings like:
```
./src/kj/common.h:1267:23: warning: initializing ‘kj::ArrayPtr<const kj::ArrayPtr<const unsigned char> >::ptr’ from ‘std::initializer_list<kj::ArrayPtr<const unsigned char> >::begin’ does not extend the lifetime of the underlying array [-Winit-list-lifetime]
```
> So not really sure I want to spend too much time looking into it, because it is very likely just buggy package.

--- a/c++/src/kj/function.h
+++ b/c++/src/kj/function.h
@@ -229,7 +229,7 @@ public:
   }
 
 private:
-  void* space[2];
+  alignas(void*) char space[2 * sizeof(void*)];
 
   class WrapperBase {
   public:

